rust/opendp/src/traits.rs

```diff
@@ -544,3 +504,328 @@ impl FloatBits for f32 {
+    }
+}
+impl_int_inf!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
+
+macro_rules! impl_float_inf_bi {
+    ($($ty:ty),+; $name:ident, $method:ident, $op:ident, $round:expr, $fallback:ident) => {
+        $(
+        #[cfg(feature="use-mpfr")]
+        impl $name for $ty {
+            fn $method(&self, other: &Self) -> Fallible<Self> {
+                use rug::float::Round::*;
+                let mut this = self.into_internal();
+                this.$op(other, $round);
+                let this = Self::from_internal(this);
+                this.is_finite().then(|| this).ok_or_else(|| err!(
+                    FailedFunction,
+                    concat!("({}).", stringify!($method), "({}) is not finite. Consider tightening your parameters."),
+                    self, other))
+            }
+        }
+        #[cfg(not(feature="use-mpfr"))]
+        impl $name for $ty {
+            fn $method(&self, other: &Self) -> Fallible<Self> {
+                let this = self.$fallback(other);
+                this.is_finite().then(|| this).ok_or_else(|| err!(
+                    FailedFunction,
+                    concat!("({}).", stringify!($method), "({}) is not finite. Consider tightening your parameters."),
+                    self, other))
+            }
+        })+
+    }
+}
+impl_float_inf_bi!(f64, f32; InfAdd, inf_add, add_assign_round, Up, alerting_add);
+impl_float_inf_bi!(f64, f32; InfSub, inf_sub, sub_assign_round, Up, alerting_sub);
+impl_float_inf_bi!(f64, f32; InfMul, inf_mul, mul_assign_round, Up, alerting_mul);
+impl_float_inf_bi!(f64, f32; InfDiv, inf_div, div_assign_round, Up, alerting_div);
+impl_float_inf_bi!(f64, f32; NegInfAdd, neg_inf_add, add_assign_round, Down, alerting_add);
+impl_float_inf_bi!(f64, f32; NegInfSub, neg_inf_sub, sub_assign_round, Down, alerting_sub);
+impl_float_inf_bi!(f64, f32; NegInfMul, neg_inf_mul, mul_assign_round, Down, alerting_mul);
+impl_float_inf_bi!(f64, f32; NegInfDiv, neg_inf_div, div_assign_round, Down, alerting_div);
+
+
+pub trait AlertingAbs: Sized {
+    fn alerting_abs(&self) -> Fallible<Self>;
+}
+macro_rules! impl_alerting_abs_signed_int {
+    ($($ty:ty),+) => ($(impl AlertingAbs for $ty {
+        fn alerting_abs(&self) -> Fallible<Self> {
+            self.checked_abs().ok_or_else(|| err!(FailedFunction,
+                "the corresponding positive value for {} is out of range", self))
+        }
+    })+)
+}
+impl_alerting_abs_signed_int!(i8, i16, i32, i64, i128);
+macro_rules! impl_alerting_abs_unsigned_int {
+    ($($ty:ty),+) => ($(impl AlertingAbs for $ty {
+        fn alerting_abs(&self) -> Fallible<Self> {
+            Ok(*self)
+        }
+    })+)
+}
+impl_alerting_abs_unsigned_int!(u8, u16, u32, u64, u128);
+macro_rules! impl_alerting_abs_float {
+    ($($ty:ty),+) => ($(impl AlertingAbs for $ty {
+        fn alerting_abs(&self) -> Fallible<Self> {
+            Ok(self.abs())
+        }
+    })+)
+}
+impl_alerting_abs_float!(f32, f64);
```

Comment:
In discussion, we proposed splitting this up, so that all traits are in a single file for clarity, and impls/macros are in a separate file.

Reply:
I did some significant refactoring here.